### ISSUE TITLE
Resolve Node export variables earlier like NodePath export variables

### ISSUE THREAD

**Issue by Spartan322** [godotengine/godot-proposals#7116]
### Describe the project you are working on

GUI-centric simulation game

### Describe the problem or limitation you are having in your project

I am actually unsure of whether to call this a bug or a feature request, it may very well be a combination of a bug discovery and a feature request:

Node export variables do not resolve early enough in tool mode if you try to use their variables for properties of other export variables, but NodePath export variables do:
```gdscript
@tool
extends Control

@export var my_node : Node

@export var rename_my_node : String:
	get: return my_node.name # runtime error: my_node is null
	set(v): my_node.name = v
```
whereas
```gdscript
@tool
extends Control

@export var my_node_path : NodePath

# This works, this will modify the node pointed to at my_node_path
# This may be unintentional as when I first save this, it crashes Godot, but on restart it always works
@export var rename_my_node : String:
	get: return get_node(my_node_path).name
	set(v): get_node(my_node_path).name = v
```

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

Making Node export variables resolve when the NodePaths are would simply make using export Node variables a lot more convenient.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

I'm not certain on the issues it could cause, but theoretically I would think that Node variables should be able to resolve in the same general place as NodePaths, or at least early enough that a property getter/setter could access it first. I suppose the one case where this might still cause problems is if you try to access a property from a Node export variable getter/setter, however this use case is already rare and resolving that would require ordering the resolution of the properties according to their contents which seems way more complicated then simply resolving Node export variables earlier.

### If this enhancement will not be used often, can it be worked around with a few lines of script?

```gdscript
@tool
extends Control

@export var my_node_path : NodePath

@export var rename_my_node : String:
	get: return get_node(my_node_path).name
	set(v): get_node(my_node_path).name = v
```

This works fine enough, aside from the fact that initially setting it may crash Godot, but it works fine and doesn't after crash more then that initial time.

### Is there a reason why this should be core and not an add-on in the asset library?

Export variables are a core feature in GDScript.

**Comment by YuriSizov** [comment]
This can't work. Nodes can only be resolved when the node is ready and thus all of its children are ready. Exported properties, however, can be get and set at any time right after the node has been initialized. You need to write your getter and setter in a safe manner and handle that case, delay execution until the node is ready (for the setter) or reject execution until that time (for the getter).

**Comment by Spartan322** [comment]
Updated my example for something that does actually work properly in my project in 4.0.3, I was thinking of what I tried beforehand which technically "works" in a broken manner. (it does modify the child nodes, but because it doesn't save the value locally it functionally does nothing) In any case the Node exports don't seem to get resolved when the NodePaths do (even though I would assume they should act generally similar) because in every case in the properties the NodePath variables will is able to reference the node, but the Node variable will always be null, the NodePath variables never log an error.

**Comment by AThousandShips** [comment]
You can't necessarily assume paths are valid at all times, this will happen for example if the node is removed from the tree, you always need to add checks and tests to ensure things are valid

**Comment by RedMser** [comment]
#325 would help here.

**Comment by YuriSizov** [comment]
> The lack of errors is why this behavior tends to confuse me, had the NodePath returned null or logged an error, I would've expected that this stayed problematic behavior.

You said it yourself that this crashes the editor for you, at least once.

**Comment by Spartan322** [comment]
> You said it yourself that this crashes the editor for you, at least once. That's a pretty huge error right there.

Yeah, I can't consistently introduce it to nodes and get it to crash, but I have noticed that sometimes upon initial introduction it does occasionally crash, that only ever happens once, and then in subsequent startups it works flawlessly without even an error, sometimes when I introduce it to nodes it doesn't crash at all and just works.
